proposals/0000-with-pushed-origin.md

```diff
@@ -0,0 +1,95 @@
+---
+Title: WithPushedOrigin instruction
+Number: 0
+Status: Draft
+Version: 0
+Authors:
+ - Francisco Aguirre
+Created: 2023-06-21
+Impact: Low
+Requires:
+Replaces:
+---
+
+## Summary
+
+The proposed change is the introduction of a `WithPushedOrigin` instruction.
+The instruction allows the XCVM to "push" a new origin for just a few instructions and then return to the original origin when those instructions are done.
+It can be seen as pushing a new origin on a stack for the execution of a block of instructions and then popping it back when the block is done.
+The new origin can only be empty (clear the origin) or a child of the current origin.
+
+The goal is to give developers more flexibility and a better experience when handling origins in their messages.
+
+## Motivation
+
+Right now, XCM has two instructions for modifying the origin: `ClearOrigin` and `DescendOrigin`.
+These work for manipulating the origins in safe ways.
+However, these instructions are final, once you use them, there's no standard way of going back to the original origin.
+This results in a complicated developer experience, where the order of operations needs to be highly taken into account to perform the operations needed, with the correct origins for each.
+
+This new instruction, `WithPushedOrigin`, provides a way of pushing an origin and popping to return to the original one.
+It makes scenarios where multiple operations need to be performed by multiple origins much easier to do.
+It gives a standard way of doing things like buying execution from a user's account and then returning to the previous origin to perform other operations that require the priviledges associated with it.
+It also allows for previously impossible scenarios like acting on behalf of many sibling origins.
+
+## Specification
```

Comment:
What would happen if the inner xcm fails? Are we going into the ErrorHandler/Appendix execution with the pushed origin or with the previous origin? 

Would also be great to see the errors that might be thrown during the execution of this instruction. Do we already have all the Errors that we need for this instruction in the format?

Reply:
I think it's good to talk about it in terms of `ClearOrigin` and `DescendOrigin` which already exist.
We would call the ErrorHandler/Appendix with the pushed origin.
No additional errors needed, all are covered by `DescendOrigin`.